execution/transaction-orderer/src/batch_orderer.rs

```diff
@@ -0,0 +1,381 @@
+
+    window: WM, // ConstOption<WindowManager<T::Key>>
+}
+
+impl<T: PTransaction> SequentialDynamicAriaOrderer<T, ConstSome<WindowManager<T::Key>>> {
+    pub fn with_window() -> Self {
+        Self {
+            txn_info: Vec::new(),
+            active_txns_count: 0,
+            selected: BTreeSet::new(),
+            write_reservations: HashMapReservationTable::default(),
+            read_reservations: HashMapReservationTable::default(),
+            window: ConstSome(WindowManager::default()),
+        }
+    }
+}
+
+impl<T: PTransaction> SequentialDynamicAriaOrderer<T, ConstNone> {
+    pub fn without_window() -> Self {
+        Self {
+            txn_info: Vec::new(),
+            active_txns_count: 0,
+            selected: BTreeSet::new(),
+            write_reservations: HashMapReservationTable::default(),
+            read_reservations: HashMapReservationTable::default(),
+            window: ConstNone(),
+        }
+    }
+}
+
+impl<T, WM> SequentialDynamicAriaOrderer<T, WM>
+where
+    T: PTransaction + Clone,
+    T::Key: Ord + Hash + Eq + Clone,
+    WM: ConstOption<WindowManager<T::Key>>,
+{
+    fn satisfy_pending_read_table_request(&mut self, idx: TxnIndex) {
+        let tx_info = &mut self.txn_info[idx];
+        assert!(tx_info.pending_read_table_requests >= 1);
+        tx_info.pending_read_table_requests -= 1;
+        if !tx_info.selected
+            && tx_info.pending_read_table_requests == 0
+            && tx_info.pending_recent_write_dependencies == 0
+        {
+            tx_info.selected = true;
+            self.selected.insert((Direction::Back, idx));
+        }
+    }
+
+    fn satisfy_pending_write_table_request(&mut self, idx: TxnIndex, key: &T::Key) {
+        let tx_info = &mut self.txn_info[idx];
+        assert!(tx_info.pending_write_table_requests >= 1);
+        tx_info.pending_write_table_requests -= 1;
+
+        match self.window.as_option_mut() {
+            Some(window) => {
+                // If window is enabled, register the dependency on the recent write.
+                // This transaction cannot be committed until this dependency is satisfied.
+                if window
+                    .recent_writes
+                    .get_mut(key)
+                    .unwrap()
+                    .dependencies
+                    .insert(idx)
+                {
+                    tx_info.pending_recent_write_dependencies += 1;
+                }
+            },
+            None => {
+                // If window is disabled and a transaction has no more pending requests
+                // in the write table, it can be selected.
+                if !tx_info.selected && tx_info.pending_write_table_requests == 0 {
+                    tx_info.selected = true;
+                    self.selected.insert((Direction::Back, idx));
+                }
+            },
+        }
+    }
+
+    fn resolve_recent_write_dependency(&mut self, idx: TxnIndex) {
```

Comment:
is the idea here that another write of the transaction is now allowed because the current set of selected transactions do not depend on it? let's add as a comment.